File: src/lib/syntax/annotations.js
/**
* @copyright 2024 - Max Bebök
* @license Apache-2.0
*/
import state from "../state.js";

export const ANNOTATIONS = {
  Barrier: "Barrier",
  Relative: "Relative",
  Align: "Align",
  NoReturn: "NoReturn",
  Unlikely: "Unlikely",
  NoRegAlloc: "NoRegAlloc",
  Tag: "Tag",
  AttrLoader: "AttrLoader",
  AttrPatch: "AttrPatch"
};

export const KNOWN_ANNOTATIONS = Object.keys(ANNOTATIONS);

/**
 * Checks if an annotation is known and valid.
 * @param {Annotation} anno
 * @throws {Error} if invalid
 */
export function validateAnnotation(anno) {
  if(!KNOWN_ANNOTATIONS.includes(anno.name)) {
    state.throwError("Unknown annotation '"+anno.name+"'!\nExpected on of: "+KNOWN_ANNOTATIONS.join(", ")+"");
  }

  // string annotations
  if(["Barrier", "AttrLoader", "AttrPatch"].includes(anno.name)) {
    if(typeof anno.value !== "string") {
      state.throwError("Annotation '"+anno.name+"' expects a string value!");
    }
    if(anno.value === "") {
      state.throwError("Annotation '"+anno.name+"' expects a non-empty string value!");
    }
  }
}

export function getAnnotationVal(annotations, name) {
  const anno = annotations.find(anno => anno.name === name);
  return anno ? (anno.value || true) : undefined;
}

File: cpp/src/annotations.cpp
#include "annotations.h"

#include "state.h"

#include <algorithm>

namespace rspl {

const std::vector<std::string> KNOWN_ANNOTATIONS = {
    "Barrier", "Relative", "Align",      "NoReturn",
    "Unlikely", "NoRegAlloc", "Tag",     "AttrLoader",
    "AttrPatch"};

// Annotations that must carry a non-empty string value.
static const std::vector<std::string> STRING_ANNOTATIONS = {
    "Barrier", "AttrLoader", "AttrPatch"};

void validateAnnotation(const std::string &name, const std::string &value,
                        const std::string &mode, bool valueIsString)
{
  if (!mode.empty()) {
    if (name != "Barrier") {
      state.throwError("Annotation @" + name + " does not take a mode argument");
    }
    if (mode != "strict" && mode != "before" && mode != "after") {
      state.throwError("Invalid @Barrier mode '" + mode +
                 "' (expected: strict, before or after)");
    }
  }

  if (std::find(KNOWN_ANNOTATIONS.begin(), KNOWN_ANNOTATIONS.end(), name) ==
      KNOWN_ANNOTATIONS.end()) {
    std::string known;
    for (size_t i = 0; i < KNOWN_ANNOTATIONS.size(); ++i) {
      if (i) known += ", ";
      known += KNOWN_ANNOTATIONS[i];
    }
    state.throwError("Unknown annotation '" + name + "'!\nExpected on of: " +
                     known);
  }

  if (std::find(STRING_ANNOTATIONS.begin(), STRING_ANNOTATIONS.end(), name) !=
      STRING_ANNOTATIONS.end()) {
    if (!valueIsString) {
      state.throwError("Annotation '" + name + "' expects a string value!");
    }
    if (value.empty()) {
      state.throwError("Annotation '" + name +
                       "' expects a non-empty string value!");
    }
  }
}

} // namespace rspl
